### PR TITLE
Handle trailing masked column behavior for nested tensor

### LINKED ISSUE #97111: TransformerEncoder truncates output when some token positions are masked by `src_key_padding_mask` across batch
### 🐛 Describe the bug

Original [post](https://discuss.pytorch.org/t/transformerencoder-src-key-padding-mask-truncates-output-when-some-token-positions-are-masked-across-batch/175316). 

**THIS ONLY HAPPENS WHEN `enable_nested_tensor=True`.**

When `src_key_padding_mask` ([N, K]) has one column (k) all with value `True`, `TransformerEncoder` will remove this column in the output, causing inconsistency issues.  Minimal example:

Class init
```python
encoder_layer = nn.TransformerEncoderLayer(d_model=256, nhead=4, dim_feedforward=512, activation='gelu', norm_first=False, batch_first=False)
self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=True)
```

Forward
```python
x = torch.randn(2000, 6, 256)
mask = torch.ones(2000, 6)
mask[:, 0] = 0  # here I masked 5 columns instead of just one
mask = mask.bool()
out = self.transformer_encoder(src = x, src_key_padding_mask = mask)
assert out.shape[1] == 6  # Error, the actual dim is only 1
```

Is this an expected behavior by design?  If so, I don't think this is a good practice because oftentimes we want aggregation on the sequence level, and this kind of removal would make the use of scatter aggregation functions, e.g. `scatter_mean` difficult.

### Versions

```
PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 9.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.83.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro RTX 8000
Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    1
Core(s) per socket:    24
Socket(s):             2
NUMA node(s):          4
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6240R CPU @ 2.40GHz
Stepping:              7
CPU MHz:               3688.476
CPU max MHz:           4000.0000
CPU min MHz:           1000.0000
BogoMIPS:              4800.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              36608K
NUMA node0 CPU(s):     0,4,8,12,16,20,24,28,32,36,40,44
NUMA node1 CPU(s):     1,5,9,13,17,21,25,29,33,37,41,45
NUMA node2 CPU(s):     2,6,10,14,18,22,26,30,34,38,42,46
NUMA node3 CPU(s):     3,7,11,15,19,23,27,31,35,39,43,47
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba rsb_ctxsw ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] denoising-diffusion-pytorch==0.31.1
[pip3] ema-pytorch==0.0.10
[pip3] numpy==1.21.5
[pip3] torch==1.13.1+cu117
[pip3] torch-cluster==1.6.1+pt113cu117
[pip3] torch-geometric==2.2.0
[pip3] torch-scatter==2.1.1+pt113cu117
[pip3] torch-sparse==0.6.17+pt113cu117
[pip3] torch-spline-conv==1.2.2+pt113cu117
[pip3] torchaudio==0.13.1+cu117
[pip3] torchdrug==0.2.0.post1
[pip3] torchvision==0.14.1+cu117
[conda] blas                      1.0                         mkl    conda-forge
[conda] cudatoolkit               10.2.89             h713d32c_10    conda-forge
[conda] denoising-diffusion-pytorch 0.31.1                   pypi_0    pypi
[conda] ema-pytorch               0.0.10                   pypi_0    pypi
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h8d4b97c_729    conda-forge
[conda] mkl-service               2.4.0            py38h95df7f1_0    conda-forge
[conda] mkl_fft                   1.3.1            py38h8666266_1    conda-forge
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.21.5           py38h6c91a56_3  
[conda] numpy-base                1.21.5           py38ha15fc14_3  
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.12.0                   pypi_0    pypi
[conda] torch-cluster             1.6.1+pt113cu117          pypi_0    pypi
[conda] torch-geometric           2.2.0                    pypi_0    pypi
[conda] torch-scatter             2.1.1+pt113cu117          pypi_0    pypi
[conda] torch-sparse              0.6.17+pt113cu117          pypi_0    pypi
[conda] torch-spline-conv         1.2.2+pt113cu117          pypi_0    pypi
[conda] torchaudio                0.13.1+cu117             pypi_0    pypi
[conda] torchdrug                 0.2.0.post1              pypi_0    pypi
[conda] torchvision               0.14.1+cu117             pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary:
Handle trailing masked column behavior for nested tensor by padding during to_padded, to original tensor size

https://github.com/pytorch/pytorch/issues/97111

Test Plan: sandcastle & github

Differential Revision: D45167874

### CODE REVIEW

**Review by bertmaher — APPROVED** [review]

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
@drisspg - can you take a look?

**Comment by cpuhrsch** [comment]
I think this introduces divergent behavior which might lead to silent correctness issues.

I suspect the issue is rather with the check that enables the use of NestedTensors and we shouldn't modify the behavior just to fix a bug.

cc @albanD @jbschlosser @jisaacso

**Comment by mikekgfb** [comment]
Belts and suspenders - yes we’re not doing it, but the updated description gives us the right to do so if we want.   Wdyt?

PS: if we do just one, which should we do?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Review by drisspg — APPROVED** [review]

File: test/test_transformers.py [hunk 1]
```diff
@@ -656,6 +656,23 @@ def test_mha_native_args(self, nb_heads, bias):
 
             mha(query=x, key=x, value=x, key_padding_mask=pad_mask)
 
+    def test_kpm_mask_trailing_column_with_nested_tensor(self):
```
Comment (drisspg) [inline test/test_transformers.py:659]: Make sure to add a `device` arg to this test and use it in the constructor of the x, and mask accordingly on rebase

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -175,6 +175,9 @@ class TransformerEncoder(Module):
             (and convert back on output). This will improve the overall performance of
             TransformerEncoder when padding rate is high. Default: ``True`` (enabled).
 
+            When enable_nested_tensor is set to True, trailing masked columns
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: The intent was to give us future freedom of action if we find that the other implementation can give us optimization opportunities.

File: test/test_transformers.py (hunk 1 above)
Comment (mikekgfb) [reply test/test_transformers.py]: Done.

File: test/test_transformers.py [hunk 3]
```diff
@@ -688,6 +688,25 @@ def test_mha_native_args(self, nb_heads, bias):
...
+    def test_kpm_mask_trailing_column_with_nested_tensor(self, device):
+        encoder_layer = nn.TransformerEncoderLayer(
+            d_model=256,
+            nhead=4,
+            dim_feedforward=512,
+            activation='gelu',
+            norm_first=False,
+            batch_first=False,
+        )
+        transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=True).to(device)
```
Comment (cpuhrsch) [inline test/test_transformers.py:700]: Is the output of

```
transformer_encoder_with_nt = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=True).to(device)
output0 = transformer_encoder(src=x, src_key_padding_mask=mask)
```

going to be the same as without nt?

```
transformer_encoder_without_nt = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=False).to(device)
output1 = transformer_encoder(src=x, src_key_padding_mask=mask)
```

Are `output0` and `output1` going to produce the same output?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
I cancelled the merge. I want to make sure that nn.TransformerEncoder behaves the same for both `enable_nested_tensor=True` and `enable_nested_tensor=False`.

**Review by cpuhrsch — APPROVED** [review]
Accepting after further offline discussion. This actually closes the gap in divergent behavior that I had asked about.
